### ISSUE TITLE
IK_Tool on MatLab API provides wrong joint angle

### ISSUE THREAD

**Issue by hkimpara** [opensim-org/opensim-core#2155]
IK_Tool on MATLAB API gave wrong joint angle with constrained joint coordinates.
On the other hand, IK_Tool on OpenSim GUI provided correct answer.

I posted an example model and script.  Please see the following Forum discussion.
[https://simtk.org/plugins/phpBB/viewtopicPhpbb.php?f=91&t=8923&p=24675](https://simtk.org/plugins/phpBB/viewtopicPhpbb.php?f=91&t=8923&p=24675)

Hopefully, this issue may be solved in soon.

Best regards, 
Hide

**Comment by tkuchida** [comment]
The results from the GUI and MATLAB are likely different because the `accuracy` property changes from `1e-05` to `1e-005` when the setup file is loaded in MATLAB (OpenSim 3.3):
<pre>
import org.opensim.modeling.*
iktool = InverseKinematicsTool('IK_Setup.xml');
iktool.print('test.xml');
</pre>
Here's the setup file: [IK_Setup.xml.txt](https://github.com/opensim-org/opensim-core/files/1945772/IK_Setup.xml.txt)

**Comment by hkimpara** [comment]
I found the accuracy became 1.0000000000000001e-05 with 4.0.

But I believe there is no numerical difference among 1e-05, 1e-005, and 1.0000000000000001e-05.  Probabry, the bugs may exist behind of these expressions.

I hope newer 4.0 can carry the fixed code.

Thank you,  
Hide

**Comment by tkuchida** [comment]
If you specify the model (*model_file*) and output (*output_motion_file*) in your setup file (lines 9 and 47), you can run it in MATLAB as follows:
<pre>
import org.opensim.modeling.*
iktool = InverseKinematicsTool('IK_Setup-2.xml');
iktool.run();
</pre>
The results I get in MATLAB are identical to those I get in the GUI (3.3). Here is the modified setup file: [IK_Setup-2.xml.txt](https://github.com/opensim-org/opensim-core/files/1946947/IK_Setup-2.xml.txt)

**Comment by hkimpara** [comment]
Hello Tom,

Thank you for finding out solutions under this condition.
But it’s really interesting feature because some methods of input model and output file name make something wrong with accuracy.

I will edit xml file directly at this moment.

Thank you so much!
Hide

**Comment by hkimpara** [comment]
Hi Tom,

Once I tried your suggested solution, but it seems to be not working as yours.
Is it different one from the latest beta release?

Thank you, 
Hide

**Comment by tkuchida** [comment]
> Is it different one from the latest beta release?

I assumed you were using 3.3 because 4.0 has not yet been released. Which version of 4.0 are you using? Did you download an installer from somewhere, or did you build locally?

**Comment by hkimpara** [comment]
Hello Tom,

I downloaded "OpenSim-4.0.Beta_032718-win64.exe" from the OpenSim web-site.
[https://simtk.org/frs/?group_id=91](url)

Even the solution script, the lumbar joint angles were quite smaller than those from OpenSim GUI (3.3).

Thank you,
Hide

**Comment by jimmyDunne** [comment]
Hi, Hide-

Thanks for sending some files so that we could reproduce your test— that was very helpful.

Using your files, I ran IK in the GUI and through Matlab and seemingly got the exact same results. 

<img width="676" alt="screen shot 2018-04-26 at 9 01 18 am" src="https://user-images.githubusercontent.com/5296507/39317429-6a4d5b8e-4930-11e8-9e1b-a81891ba2fcb.png">

I performed IK in Matlab with the following code;
<pre>
import org.opensim.modeling.*
model = Model('Lumbar_C_4_HK.osim');
iktool = InverseKinematicsTool('IK_Setup-2.xml');
iktool.setModel(model)
iktool.run();
</pre>

In your code, you perform some other operations (like `model.assemble(s);`) before executing. Perhaps those calls are having conflicting effects on the tool. It seems like Tom and I have both been able to get the same results from either Matlab or GUI.

**Comment by hkimpara** [comment]
Hello James,

Thank you for working on this example.
I think the magnitude of lumbar joint angles are quite small.  

Please see following images.  The GUI based IK showed zigzag shaped joint angle results.

The flex_extension must reach to -20 degrees at 0.2 s.  However, MATLAB API result shows only -0.35 degrees as you showed here.
![lumbarjointangles](https://user-images.githubusercontent.com/19690418/39324669-0fa98bc0-495e-11e8-9262-bdecf7f0c929.png)

However, Hip joint angles are OK.
![hipjointangles](https://user-images.githubusercontent.com/19690418/39324824-73065db0-495e-11e8-8794-7a70279e5dee.png)

If you see the lumbar joint angles less than -1 degrees in GUI based IK, it's also wrong result.

Thank you,
Hide

**Comment by hkimpara** [comment]
I'd like to share more clear images of this example.  I added target points such as spherical bodies with red or blue colors as Lumbar_C_4_HK2.osim.  Those locations are exact same as "Marker_Points.trc".  Please see the shoulder markers.  These markers would pass through the blue colored points.  But MATLAB IK results did not pass through these points.

[Ideal_postures.pdf](https://github.com/opensim-org/opensim-core/files/1952853/Ideal_postures.pdf)

[Lumbar_C_4_HK2.osim.txt](https://github.com/opensim-org/opensim-core/files/1952858/Lumbar_C_4_HK2.osim.txt)

**Comment by hkimpara** [comment]
Hello Tom and James,

I realized that the coordinate definitions for lumbar joint is Coupled, whereas those for hip joint is Rotational.  Then in OpenSim 4.0, Rotational coordinates are expressed in degrees but Coupled coordinates are expressed in radians.  On the other hand, OpenSim 3.3 expressed Coupled coordinates in degrees.  Is it correct?
I keep using GUI 3.3 with 4.0 result because of better graphic views in 3.3 GUI.  But I may need to switch GUI to 4.0 for 4.0 results.

If my understanding is correct, it's not bug issue any more.

Thank you,
Hide

**Comment by jimmyDunne** [comment]
I spent some time running your files through the 4.0 GUI beta and Matlab API. At the end of it, I still got the exact same results for the GUI and the Matlab API. There were some issues with the GUI when viewing the .trc files— I am not sure if that affected your results, but I doubt it.

I have attached a zip with all the files I used and the results. You will see that the results files are the same. 

[Report04_IK_Const_james.zip](https://github.com/opensim-org/opensim-core/files/1956104/Report04_IK_Const_james.zip)

**Comment by hkimpara** [comment]
Hello Jim,

Thank you for spending your time for my case.  I think "Coupled" coordinate unit has been changed from degrees with 3.3 to radian with 4.0.  When I open my model in 3.3, the unit of coordinate for lumbar joint was expressed in degrees.  On the other hand, in 4.0, the unit looks like radian.  

I agree to close this issue at this moment. 
However, please clarify the unit difference of coupled coordinate in change log or doxygen manual.  Otherwise users may sometimes misunderstand the magnitude state variables as me.

Best regards,
Hide

**Comment by tkuchida** [comment]
Thanks, Hide. I thought this issue might be related to #2065, but I don't see any LinearFunctions in your model with a slope of -1, so I don't know. We should find out why the Coordinates panel in the GUI uses degrees in 3.3 and radians in 4.0, and either add a note to CHANGELOG.md as  you suggest, or make the GUI consistent with how the API treats these Coordinates.

**Coordinates panel in 3.3:**
![](https://user-images.githubusercontent.com/4203505/39380683-435195f6-4a14-11e8-9cf7-9ed2bf79e004.png)

**Coordinates panel in 4.0 (AppVeyor artifact OpenSim-e5f97340-2018-04-20.zip):**
![](https://user-images.githubusercontent.com/4203505/39380687-4681130a-4a14-11e8-9784-6c0cd0e1dc7d.png)

@aymanhab How does the GUI determine the MotionType in this case?
Model file: [Lumbar_C_4_HK.osim.txt](https://github.com/opensim-org/opensim-core/files/1956363/Lumbar_C_4_HK.osim.txt)

**Comment by hkimpara** [comment]
Hi Tom,

The ikTool provides results with isDegrees=yes in both of 3.3 and 4.0.  As the results, joint angles with coupled coordinate are expressed in degree from 3.3 but in radian from 4.0.  I got a trouble because I opened the result file by 3.3 GUI without recognizing this difference.  But if the ikTool provides results with isDegrees=no and all values in radian, I may not be confused at that moment.

But I believe the definition and usage of units are very important for end users.  I hope no other users may not be confused like me.

Best regards,
Hide
